File: Add Two Numbers.cpp
//fail
//상당히 깔끔한 링크드 리스트 코드이다.
//문제:
//두개의 0이상의 정수를 나타내는, 비어있지 않은 링크드 리스트 두 개가 주어진다. 이 숫자들은 역순으로 저장되어있고, 
//각 노드들은 하나의 숫자를 갖고 있따.
//두 링크드 리스트를 더해라.

//해결 아이디어:
//dummyNode는 리턴하기 위한 용도, curr은 while문에서 리스트를 순회하기 위한 용도,
//그리고 코드를 복잡하게 하지 않기 위해, while문의 조건과 그 안의 if문 조건들을 잘 보자.
//구현 자체는 간단하나 코드의 간결성을 보고 넘어가자.

//실수한점?
//마지막에 upper가 발생하는데, 새로운 노드를 추가해 줘야 할 경우 생각해주지않았다.
//기존에 있던 리스트에다가 더하는 방식은 오히려 새로운 예외를 나을 수도 있다.


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummyHead = new ListNode(0);
        ListNode* p = l1;
        ListNode* q = l2;
        ListNode* curr = dummyHead;
        int carry = 0;
        
        while(p != NULL || q != NULL){
            int x = (p != NULL) ? p->val : 0;
            int y = (q != NULL) ? q->val : 0;
            int sum = carry + x + y;
            carry = sum / 10;
            curr->next = new ListNode(sum % 10);
            curr = curr->next;
            if(p != NULL) p = p->next;
            if(q != NULL) q = q->next;
        }
        if(carry > 0)
            curr->next = new ListNode(carry);
        
        return dummyHead->next;
    }
};

File: kotlin/Add Two Numbers.kt
class Solution {
    /*
    * 코틀린은 기본적으로 값이 null이 아니다. 기본적으로는 NotNull이다.
    * Nullable 표현에만 '?'가 사용된다.
    *
    * null을 처리하는 방법 중에 엘비스 연산자(?:)가 있다.
    * p1?.`val` ?: 0 의 뜻은 nullable일수 있는 p1의 val이란 변수가 null일 경우 0으로 대체한다 라는 뜻이다.
    * 참고: https://ddolcat.tistory.com/618
    * */
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val fakeHead = ListNode(0)
        var cur = fakeHead
        var p1 = l1
        var p2 = l2
        var carry = 0
        while(p1 != null || p2 != null || carry != 0){
            val curSum = (p1?.`val` ?: 0) + (p2?.`val` ?: 0) + carry
            carry = curSum / 10
            cur.next = ListNode(curSum % 10)
            cur = cur.next!!
            p1 = p1?.next
            p2 = p2?.next
        }
        return fakeHead.next
    }
}
